// @formio/premium/lib/components/DataTable/customFilters.js
import * as _ from 'lodash';
var selectFilters = [
    {
        type: 'equal',
        label: 'Is equal to',
        filter: function (value, filter) { return _.isEqual(value, filter); }
    },
    {
        type: 'notEqual',
        label: 'Is not equal to',
        filter: function (value, filter) { return !_.isEqual(value, filter); }
    },
    {
        type: 'contain',
        label: 'Contains',
        filter: function (value, filter) {
            if (_.isArray(value)) {
                return filter.every(function (f) { return value.includes(f); });
            }
            return _.isEqual(value, filter);
        }
    },
    {
        type: 'notContain',
        label: 'Does not contain',
        filter: function (value, filter) {
            if (_.isArray(value)) {
                return filter.every(function (f) { return !value.includes(f); });
            }
            return !_.isEqual(value, filter);
        }
    },
];
var checkboxFilters = [
    {
        type: 'equal',
        label: 'Is equal to',
        filter: function (value, filter) { return _.isEqual(value, filter); }
    }
];
var selectboxesFilters = [
    {
        type: 'equal',
        label: 'Is equal to',
        filter: function (value, filter) { return _.isEqual(value, filter); }
    },
    {
        type: 'contain',
        label: 'Contains',
        filter: function (value, filter) { return value.toString().includes(filter.toString()); }
    },
];
var dayFilters = [
    {
        type: 'equal',
        label: 'Is equal to',
        filter: function (value, filter) { return _.isEqual(value, filter); }
    },
    {
        type: 'notEqual',
        label: 'Is not equal to',
        filter: function (value, filter) { return !_.isEqual(value, filter); }
    },
    {
        type: 'gte',
        label: 'Is after',
        filter: function (value, filter) { return value >= filter; }
    },
    {
        type: 'lte',
        label: 'Is before',
        filter: function (value, filter) { return value <= filter; }
    },
];
var surveyFilters = [
    {
        type: 'contain',
        label: 'Contains',
        filter: function (value, filter) { return value.toString().includes(filter.toString()); }
    }
];
var signatureFilters = [
    {
        type: 'exists',
        label: 'Is equal to',
        filter: function (value, filter) { return _.isEqual(!!value, filter); }
    }
];
var fileFilters = [
    {
        type: 'exists',
        label: 'Uploaded',
        filter: function (value, filter) { return _.isEqual(!!value, filter); }
    }
];
export var filterByComponentType = {
    select: selectFilters,
    checkbox: checkboxFilters,
    selectboxes: selectboxesFilters,
    day: dayFilters,
    survey: surveyFilters,
    signature: signatureFilters,
    file: fileFilters
};
