// @formio/premium/lib/modules/RulesEngine/components/Component/editForm/rulesTabs/Rules.edit.conditionalVisibility.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import EditFormUtils from '../utils';
import BaseConditional from 'formiojs/components/_classes/component/editForm/Component.edit.conditional';
export default [
    {
        label: 'Conditional visibility:',
        tooltip: 'Select the complexity of visibility condition',
        optionsLabelPosition: 'right',
        inline: true,
        values: [
            { label: 'Simple', value: 'simple', shortcut: '' },
            { label: 'Advanced', value: 'advanced', shortcut: '' },
        ],
        key: 'conditionalVisibilityComplexity',
        type: 'radio',
        input: true,
        weight: 0,
    },
    __assign(__assign({}, BaseConditional.find(function (setting) { return setting.key === 'simple-conditional'; })), { type: 'custom', hideLabel: true, conditional: {
            show: true,
            when: 'conditionalVisibilityComplexity',
            eq: 'simple',
        }, weight: 10, input: false }),
    {
        label: 'Advanced Condition',
        key: 'advancedConditionPanel',
        hideLabel: true,
        type: 'custom',
        input: false,
        weight: 10,
        conditional: {
            show: true,
            when: 'conditionalVisibilityComplexity',
            eq: 'advanced',
        },
        components: __spreadArrays(EditFormUtils.getModesComponents('Create visibility condition using:', 'advancedVisibilityCondition', 0, {
            ui: EditFormUtils.advancedConditionEditFormComponents.ui(),
            js: EditFormUtils.advancedConditionEditFormComponents.js(),
            json: EditFormUtils.advancedConditionEditFormComponents.json(),
        }))
    },
];
