// @formio/premium/lib/components/DataTable/editForm/DataTable.edit.fetch.js
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import PremiumUtils from '../../../utils';
export default [
    {
        type: 'checkbox',
        input: true,
        weight: 0,
        key: 'fetch.enableFetch',
        label: 'Enable fetch',
        defaultValue: false,
        tooltip: 'If fetch is enabled, the DataTable will be automatically populated with data from the data source and will not be submitted.',
    },
    {
        type: 'panel',
        title: 'Fetch Settings',
        key: 'fetch',
        weight: 700,
        customConditional: function (_a) {
            var data = _a.data, _ = _a._;
            return _.get(data, 'fetch.enableFetch', false);
        },
        components: __spreadArrays(PremiumUtils.baseEditFetchComponents([
            {
                key: 'dataSrc',
                ignore: true,
            },
            {
                key: 'indexeddb.fromExisting',
                ignore: true,
            },
            {
                key: 'indexeddb.database',
                ignore: true,
            },
            {
                key: 'indexeddb.table',
                ignore: true,
            },
            {
                key: 'indexeddb.database',
                ignore: true,
            },
            {
                key: 'indexeddb.table',
                ignore: true,
            },
            {
                key: 'fetch.url',
                conditional: '',
            },
            {
                key: 'fetch.method',
                ignore: true,
            },
            {
                key: 'fetch.headers',
                conditional: ''
            },
            {
                key: 'fetch.forwardHeaders',
                ignore: true,
            },
            {
                key: 'fetch.authenticate',
                conditional: ''
            },
            {
                key: 'specifyPostBodyPanel',
                ignore: true,
            },
            {
                key: 'allowCaching',
                conditional: ''
            },
        ]), [
            {
                type: 'textfield',
                input: true,
                key: 'fetch.provider',
                weight: 10,
                label: 'Custom Fetch Provider',
                placeholder: 'Enter the name of your fetch provider (optional)',
                tooltip: 'A name of custom fetch provider. If not specified, default URL provider is used.',
                clearOnHide: true,
            },
            {
                label: 'Default Sort Query',
                tooltip: 'This sort query is used if the data table culumn sorting is not set.',
                tableView: true,
                weight: 20,
                key: 'fetch.sort.defaultQuery',
                type: 'textfield',
                input: true
            },
            {
                label: 'Descending Order',
                tooltip: 'Type of default sorting.',
                tableView: false,
                defaultValue: false,
                weight: 30,
                key: 'fetch.sort.descending',
                customConditional: 'show = !!_.get(data, "fetch.sort.defaultQuery", "");',
                type: 'checkbox',
                input: true
            },
        ])
    },
];
