// @formio/premium/lib/modules/RulesEngine/validators/transformers/Property.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import _ from 'lodash';
import { Transformer } from './Transformer';
var PropertyTransformer = /** @class */ (function (_super) {
    __extends(PropertyTransformer, _super);
    function PropertyTransformer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(PropertyTransformer, "title", {
        get: function () {
            return 'Property';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PropertyTransformer, "itemName", {
        get: function () {
            return 'property';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PropertyTransformer, "itemArguments", {
        get: function () {
            return [
                {
                    name: 'Path',
                    key: 'path',
                    required: false,
                    extendedContainerParameters: {
                        tooltip: 'Specify path to property',
                    }
                },
                {
                    name: 'Default Value',
                    key: 'defaultValue',
                    extendedContainerParameters: {
                        tooltip: 'Specify value that will be used when property is undefined',
                    },
                    required: false,
                },
            ];
        },
        enumerable: false,
        configurable: true
    });
    PropertyTransformer.prototype.transform = function (value, args) {
        var path = args.path, defaultValue = args.defaultValue;
        return _.get(value, path, defaultValue !== null && defaultValue !== void 0 ? defaultValue : value);
    };
    return PropertyTransformer;
}(Transformer));
export { PropertyTransformer };
