// @formio/premium/lib/components/Radio/Radio.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Components } from 'formiojs';
var RadioComponent = Components.components.radio;
var initialSimpleConditionSettings = RadioComponent.simpleConditionSettings;
Object.defineProperty(RadioComponent, 'simpleConditionSettings', {
    enumerable: true,
    configurable: true,
    get: function () {
        return __assign(__assign({}, initialSimpleConditionSettings), { valueComponent: function (classComp) {
                return {
                    type: 'select',
                    dataSrc: 'custom',
                    valueProperty: 'value',
                    dataType: classComp.dataType || '',
                    data: {
                        custom: function () {
                            return classComp.values;
                        }
                    },
                };
            } });
    }
});
export default RadioComponent;
