// @formio/premium/lib/components/SignRequestSignature/editForm/SignRequestSignature.edit.data.js
export default [
    {
        label: 'Order',
        key: 'order',
        input: true,
        type: 'textfield',
        tooltip: 'The order in which the signers will receive emails. Zero (0) is for the owner of the document.',
        validate: {
            required: true,
        },
        weight: 0
    },
    {
        label: 'Signer Email',
        key: 'email',
        input: true,
        type: 'textfield',
        tooltip: 'An email of a person that needs to sign the document.',
        clearOnHide: true,
        validate: {
            required: true,
            custom: 'valid = /\\{\\{([\\s\\S]+?)\\}\\}/g.test(input) ? true : /^(([^<>()[\\]\\\\.,;:\\s@\"]+(\\.[^<>()[\\]\\\\.,;:\\s@\"]+)*)|(\".+\"))@((\\[[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$/.test(input)',
            customMessage: 'Must be a valid email'
        },
        conditional: {
            show: false,
            when: 'order',
            eq: 0
        },
        weight: 5
    },
    {
        label: 'Signer Name',
        key: 'name',
        input: true,
        type: 'textfield',
        tooltip: 'A name that will be written on a signature by default.',
        weight: 10,
    },
    {
        key: 'multiple',
        ignore: true,
    },
    {
        key: 'defaultValue',
        ignore: true,
    },
    {
        key: 'clearOnHide',
        ignore: true,
    },
];
