// @formio/premium/lib/modules/RulesEngine/validators/quickRules/Required.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import _ from 'lodash';
import { QuickRule } from './QuickRule';
var RequiredQuickRule = /** @class */ (function (_super) {
    __extends(RequiredQuickRule, _super);
    function RequiredQuickRule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(RequiredQuickRule, "itemName", {
        get: function () {
            return 'required';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RequiredQuickRule, "title", {
        get: function () {
            return 'Required';
        },
        enumerable: false,
        configurable: true
    });
    RequiredQuickRule.prototype.addRule = function (helper) {
        var conditionName = 'Required';
        var conditionKey = helper.makeConditonKeyUnique(_.camelCase(conditionName));
        helper.addCondition({
            name: conditionName,
            key: conditionKey,
            parts: [
                {
                    type: 'new',
                    value: {
                        valueSource: 'thisComponent'
                    },
                    operator: {
                        name: 'notIsEmpty',
                        notIsEmptyArguments: {},
                    },
                },
            ],
        });
        var validation = {
            condition: conditionKey,
            conditionType: 'new',
            severity: 'error',
            message: 'Component is required',
        };
        helper.addValidation(validation);
        helper.execute();
        return validation;
    };
    return RequiredQuickRule;
}(QuickRule));
export { RequiredQuickRule };
