// @formio/premium/lib/providers/storage/azure.js
import XHR from './xhr';
var Azure = /** @class */ (function () {
    function Azure(formio) {
        this.formio = formio;
    }
    Azure.prototype.uploadFile = function (file, fileName, dir, progressCallback, url, options, fileKey, groupPermissions, groupId) {
        return XHR.upload(this.formio, 'azure', function (xhr, response) {
            xhr.open('PUT', response.url);
            xhr.setRequestHeader('Content-Type', file.type);
            xhr.setRequestHeader('x-ms-blob-type', 'BlockBlob');
            return file;
        }, file, fileName, dir, progressCallback, groupPermissions, groupId).then(function () {
            return {
                storage: 'azure',
                name: XHR.path([dir, fileName]),
                size: file.size,
                type: file.type,
                groupPermissions: groupPermissions,
                groupId: groupId
            };
        });
    };
    Azure.prototype.downloadFile = function (file) {
        return this.formio.makeRequest('file', this.formio.formUrl + "/storage/azure?name=" + XHR.trim(file.name), 'GET');
    };
    Azure.title = 'Azure File Services';
    return Azure;
}());
export default Azure;
