// @formio/premium/lib/modules/RulesEngine/index.js
import { Formio } from 'formiojs';
import { BaseEntity } from './validators/BaseEntity';
import { BaseCalculatableEntity } from './validators/BaseCalculatableEntity';
import { Conjunction } from './validators/conjunctions/Conjunction';
import { QuickRule } from './validators/quickRules/QuickRule';
import { Operator } from './validators/operators/Operator';
import { BaseReduceTransformer } from './validators/transformers/BaseReduce';
import { IterateeTransformer } from './validators/transformers/Iteratee';
import { Transformer } from './validators/transformers/Transformer';
import { ValueSource } from './validators/valueSources/ValueSource';
import conjunctions from './validators/conjunctions';
import operators from './validators/operators';
import quickRules from './validators/quickRules';
import transformers from './validators/transformers';
import valueSources from './validators/valueSources';
import components from './components';
import builders from './builders/webform/WebformBuilder';
import templates from './templates';
export * from './utils/utils';
export var Abstract = {
    BaseCalculatableEntity: BaseCalculatableEntity,
    BaseEntity: BaseEntity,
    BaseReduceTransformer: BaseReduceTransformer,
    Conjunction: Conjunction,
    IterateeTransformer: IterateeTransformer,
    QuickRule: QuickRule,
    Operator: Operator,
    Transformer: Transformer,
    ValueSource: ValueSource,
};
var rulesEngine = Formio.rulesEntities
    ? {
        conjunctions: conjunctions,
        operators: operators,
        quickRules: quickRules,
        transformers: transformers,
        valueSources: valueSources,
        components: components,
        builders: builders,
        templates: templates
    }
    : {};
export default rulesEngine;
