// @formio/premium/lib/components/DynamicWizard/DynamicWizard.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import { Formio, Components, Utils } from 'formiojs';
var EditGrid = Components.components.editgrid;
var NestedComponent = Components.components.nested;
import editForm from './DynamicWizard.form';
import templates from './templates';
import _ from 'lodash';
var FormState = {
    New: 'new',
    Editing: 'editing',
    Saved: 'saved',
    Removed: 'removed',
    Draft: 'draft',
};
var DynamicWizard = /** @class */ (function (_super) {
    __extends(DynamicWizard, _super);
    function DynamicWizard() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = _super.apply(this, args) || this;
        _this.type = 'dynamicWizard';
        _this.isChangingMode = true;
        _this.isSavingInProgress = false;
        _this.editRows = [];
        _this.changingRowIndex = null;
        _this.step = 0;
        _this.buttonSettings = {
            showPrevious: true,
            showNext: true,
            showSubmit: true,
            showCancel: !_this.options.readOnly
        };
        _this.shouldUpdate = true;
        return _this;
    }
    DynamicWizard.schema = function () {
        var extend = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            extend[_i] = arguments[_i];
        }
        return EditGrid.schema.apply(EditGrid, __spreadArrays([{
                type: 'dynamicWizard',
                label: 'Dynamic Wizard',
                key: 'dynamicWizard',
                templates: {
                    header: DynamicWizard.defaultHeaderTemplate,
                    row: DynamicWizard.defaultRowTemplate,
                    footer: '',
                },
            }], extend));
    };
    Object.defineProperty(DynamicWizard, "builderInfo", {
        get: function () {
            return {
                title: 'Dynamic Wizard',
                group: 'premium',
                icon: 'tasks',
                ignoreForForm: true,
                disableSiblings: true,
                weight: 20,
                schema: DynamicWizard.schema(),
            };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard, "defaultHeaderTemplate", {
        get: function () {
            return "<div class=\"row\" role=\"row\">\n    <div class=\"col-sm-2\">\n      Users\n    </div>\n  </div>";
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard, "defaultRowTemplate", {
        get: function () {
            return "<div class=\"list-group-item\">\n    <div class=\"list-group-subheader\">\n      <div class=\"row\" role=\"row\">\n        <div class=\"col-sm-2\">\n          User {{ rowIndex + 1 }}\n        </div>\n        {% if (!ctx.self.options.readOnly && !component.disabled) { %}\n        <div class=\"col-sm-2\" role=\"gridcell\">\n          <div class=\"btn-group pull-right\">\n            <button class=\"btn btn-default btn-light btn-sm editCard\" aria-label=\"{{ ctx.t('Edit row ' + (rowIndex + 1)) }}\"><i class=\"{{ ctx.iconClass('edit') }}\"></i></button>\n            {% if (!instance.hasRemoveButtons || instance.hasRemoveButtons()) { %}\n            <button class=\"btn btn-danger btn-sm removeRow\" aria-label=\"{{ ctx.t('Remove row ' + (rowIndex + 1)) }}\"><i class=\"{{ ctx.iconClass('trash') }}\"></i></button>\n            {% } %}\n          </div>\n        </div>\n        {% } %}\n      </div>\n    </div>\n    {% ctx.util.eachComponent(ctx.components, function(component) { %}\n    {% if ((!component.hasOwnProperty('tableView') || component.tableView) && isVisibleInRow(component)) { %}\n    <div class=\"row\" role=\"row\">\n      <div class=\"col-sm-2\" role=\"gridcell\">\n        {{ component.key }}\n      </div>\n    </div>\n    <div class=\"row\" role=\"row\">\n      <div class=\"col-sm-2\" role=\"gridcell\">\n        {{ ctx.getView(component, ctx.row[component.key]) }}\n      </div>\n    </div>\n    {% } %}\n    {% }) %}\n  </div>";
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "dynamicWizardKey", {
        get: function () {
            return "dynamicWizard-" + this.key;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "rowRef", {
        get: function () {
            return this.dynamicWizardKey + "-row";
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "rowElements", {
        get: function () {
            return this.refs[this.rowRef];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "rowRefs", {
        get: function () {
            return this.refs["dynamicWizard-" + this.component.key + "-row"];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "agreeButtonRef", {
        get: function () {
            return this.dynamicWizardKey + "-agreeButton";
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "denyButtonRef", {
        get: function () {
            return this.dynamicWizardKey + "-denyButton";
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "agreeButtonElements", {
        get: function () {
            return this.refs[this.agreeButtonRef];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "cancelRowRef", {
        get: function () {
            return this.dynamicWizardKey + "-cancelRow";
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "cancelRowElements", {
        get: function () {
            return this.refs[this.cancelRowRef];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "inlineEditMode", {
        get: function () {
            return this.component.inlineEdit;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "saveEditMode", {
        get: function () {
            return !this.inlineEditMode;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "rootWizard", {
        get: function () {
            var _a;
            return this.findRootWizard((_a = this.parent) === null || _a === void 0 ? void 0 : _a.parent) || {};
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "hasComponents", {
        get: function () {
            return this.component.components.length;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "buttons", {
        get: function () {
            var _this = this;
            var buttons = {};
            [
                { name: 'cancel', method: 'cancelRow' },
                { name: 'previous', method: 'prevPage' },
                { name: 'next', method: 'nextPage' },
            ].forEach(function (button) {
                if (_this.hasButton(button.name)) {
                    buttons[button.name] = button;
                }
            });
            return buttons;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DynamicWizard.prototype, "defaultSchema", {
        get: function () {
            return DynamicWizard.schema();
        },
        enumerable: false,
        configurable: true
    });
    DynamicWizard.prototype.init = function (secondRender, prevVisibility) {
        if (prevVisibility === void 0) { prevVisibility = this._visible; }
        this.prevVisibility = prevVisibility;
        this.prevBlocking = this.disabled;
        this.secondRender = !!secondRender;
        this.shouldUpdate = true;
        this.changingRowIndex = null;
        if (this.builderMode) {
            this.editRows = [];
        }
        return _super.prototype.init.call(this);
    };
    DynamicWizard.prototype.render = function (children) {
        var _a, _b, _c;
        if (this.builderMode) {
            return _super.prototype.render.call(this);
        }
        var dataValue = this.dataValue || [];
        var headerTemplate = Utils.Evaluator.noeval ? templates.header : Utils._.get(this.component, 'templates.header');
        var editRow = this.editRows[this.changingRowIndex];
        if (this.step === 0 && editRow && !editRow.components[this.step]._visible) {
            this.nextPage();
        }
        var reset = !this.editRows.length && this.step === 0 && this.changingRowIndex === 0;
        if (((!this.secondRender && this.shouldUpdate) || reset) && !this.options.readOnly) {
            this.secondRender = true;
            if (!this.rootWizard.editMode) {
                var onFirstPage = this.root.page === 0 && this.page === this.root.page;
                if (onFirstPage && this._parentVisible && this._visible && !this._disabled && this.hasComponents) {
                    this.shouldUpdate = false;
                    this.isInitRowExists = true;
                    this.addRow(true, reset);
                }
            }
        }
        if (((_a = this.root) === null || _a === void 0 ? void 0 : _a.editMode) && this.rootWizard.editMode && this.isInitRowExists) {
            this.isInitRowExists = false;
            this.cancelRow();
        }
        return _super.prototype.render.call(this, children || this.renderTemplate('dynamicWizard', {
            ref: {
                row: this.rowRef,
                agreeButton: this.agreeButtonRef,
                denyButton: this.denyButtonRef,
            },
            header: this.renderString(headerTemplate),
            footer: this.renderString(Utils._.get(this.component, 'templates.footer'), {
                components: this.component.components,
                value: dataValue,
            }),
            rows: this.editRows.map(this.renderRow.bind(this)),
            currentComponent: this.step === -1 ? null : (_c = (_b = this.editRows[this.changingRowIndex]) === null || _b === void 0 ? void 0 : _b.components[this.step]) === null || _c === void 0 ? void 0 : _c.render(),
            errors: this.editRows.map(function (row) { return row.error; }),
            buttons: this.buttons,
            hasAddButton: this.hasAddButton(),
            hasRemoveButtons: this.hasRemoveButtons(),
            isChangingMode: this.isChangingMode,
            isDisabled: this._disabled,
            isBlocking: !this._visible || this._disabled || !this.hasComponents,
            dynamicWizardKey: this.dynamicWizardKey,
            readOnly: this.options.readOnly,
        }));
    };
    DynamicWizard.prototype.attach = function (element) {
        var _a;
        var _this = this;
        var superAttach = _super.prototype.attach.call(this, element);
        if (this.builderMode) {
            return _super.prototype.attach.call(this, element);
        }
        this.visibilityCheck();
        this.blockingCheck();
        this.loadRefs(element, (_a = {},
            _a[this.agreeButtonRef] = 'multiple',
            _a[this.denyButtonRef] = 'multiple',
            _a[this.rowRef] = 'multiple',
            _a[this.dynamicWizardKey + "-cancel"] = 'single',
            _a[this.dynamicWizardKey + "-previous"] = 'single',
            _a[this.dynamicWizardKey + "-next"] = 'single',
            _a));
        this.agreeButtonElements.forEach(function (agreeButton) {
            _this.addEventListener(agreeButton, 'click', function () {
                _this.addRow();
                if (_this.rootWizard.currentNextPage === -1) {
                    _this.rootWizard.redraw();
                }
            });
        });
        this.attachNav();
        if (!this.isChangingMode) {
            this.rowElements.forEach(function (row, rowIndex) {
                // Attach edit and remove button events.
                [
                    {
                        className: 'removeCard',
                        event: 'click',
                        action: function () { return _this.removeRow(rowIndex); },
                    },
                    {
                        className: 'editCard',
                        event: 'click',
                        action: function () {
                            _this.editRow(rowIndex);
                        },
                    },
                ].forEach(function (_a) {
                    var className = _a.className, event = _a.event, action = _a.action;
                    var elements = row.getElementsByClassName(className);
                    Array.prototype.forEach.call(elements, function (element) {
                        _this.addEventListener(element, event, action);
                    });
                });
            });
        }
        else {
            var editRow = this.editRows[this.changingRowIndex];
            if (editRow) {
                this.attachComponents(this.rowElements[0], editRow.components, this.component.components, true);
            }
        }
        return superAttach;
    };
    DynamicWizard.prototype.attachNav = function () {
        var _this = this;
        Utils._.each(this.buttons, function (button) {
            var buttonElement = _this.refs[_this.dynamicWizardKey + "-" + button.name];
            _this.addEventListener(buttonElement, 'click', function (event) {
                event.preventDefault();
                var cancelRedraw = _this[button.method]();
                if (cancelRedraw && button.method === 'nextPage') {
                    return;
                }
                _this.redraw().then(function () {
                    _this.afterAttachNav(button.method);
                });
                var editRow = _this.editRows[_this.changingRowIndex];
                if (editRow) {
                    _this.validateRow(editRow, false);
                }
            });
        });
    };
    DynamicWizard.prototype.afterAttachNav = function (method) {
        var _a;
        var editRow = this.editRows[this.changingRowIndex];
        switch (method) {
            case 'prevPage':
            case 'nextPage':
                if (this.step === -1) {
                    this.agreeButtonElements[0].focus();
                }
                else {
                    var input = (_a = editRow.components[this.step].refs) === null || _a === void 0 ? void 0 : _a.input;
                    input ? input[0].focus() : this.focusOnNewRowElement(editRow.components, this.step);
                }
        }
    };
    DynamicWizard.prototype.attachComponents = function (element, components, container, isDynamicWizard) {
        var _a;
        if (container === void 0) { container = this.component.components; }
        if (this.builderMode) {
            return _super.prototype.attachComponents.call(this, element, components, container);
        }
        if (!isDynamicWizard) {
            return;
        }
        components = components || this.component.components;
        element = this.hook('attachComponents', element, components, container, this);
        if (!element) {
            // Return a non-resolving promise.
            return (new Formio.Promise(function () {
                // empty
            }));
        }
        var index = 0;
        var promises = [];
        if (!this.isChangingMode) {
            Array.prototype.slice.call(element.children).forEach(function (child) {
                if (!child.getAttribute('data-noattach') && components[index]) {
                    promises.push(components[index].attach(child));
                    index++;
                }
            });
        }
        else if (!Utils._.isNil(this.step) && element.children[0] && !((_a = element.children[0]) === null || _a === void 0 ? void 0 : _a.getAttribute('data-noattach')) && components[this.step]) {
            promises.push(components[this.step].attach(element.children[0]));
        }
        return Formio.Promise.all(promises);
    };
    DynamicWizard.prototype.visibilityCheck = function () {
        if (this.page === this.root.page) {
            if (this._visible && !this.prevVisibility) {
                this.prevVisibility = true;
                if (!this.disabled && this.hasComponents) {
                    this.addRow();
                }
            }
            if (this.prevVisibility && !this._visible) {
                this.prevVisibility = false;
            }
        }
    };
    DynamicWizard.prototype.blockingCheck = function () {
        if (this.page === this.root.page && this.prevBlocking && !this.disabled) {
            this.prevBlocking = false;
            this.isChangingMode = false;
            this.rootWizard.redraw();
        }
    };
    DynamicWizard.prototype.isOpen = function (editRow) {
        return [FormState.New, FormState.Editing].includes(editRow === null || editRow === void 0 ? void 0 : editRow.state);
    };
    DynamicWizard.prototype.resetValue = function () {
        NestedComponent.prototype.resetValue.call(this);
        this.emptyRows();
        this.step = 0;
        this.addRow(true, true);
    };
    DynamicWizard.prototype.addRow = function (firstPage, reset) {
        if (firstPage === void 0) { firstPage = false; }
        if (reset === void 0) { reset = false; }
        if (this.options.readOnly) {
            return;
        }
        if (!this.hasAddButton() && !this.isOpen(this.editRows[this.changingRowIndex])) {
            if (this.isChangingMode) {
                this.isChangingMode = false;
            }
            this.rootWizard.element.classList.remove('dynamicWizard-changingMode');
            this.rootWizard.redraw();
            return;
        }
        var dataObj = {};
        var rowIndex = this.editRows.length;
        var editRow;
        if (Utils._.isNil(this.changingRowIndex) || reset) {
            this.changingRowIndex = this.editRows.length;
            editRow = {
                components: this.createRowComponents(dataObj, rowIndex),
                data: dataObj,
                state: FormState.New,
                backup: null,
                error: null,
            };
            this.editRows.push(editRow);
            if (this.inlineEditMode) {
                this.dataValue.push(dataObj);
                this.triggerChange();
            }
            this.checkRow('checkData', null, {}, editRow.data, editRow.components);
            if (this.component.modal) {
                this.addRowModal(rowIndex);
            }
        }
        this.isChangingMode = true;
        this.step = this.step !== -1 ? this.step : 0;
        this.addChangingMode(firstPage);
        return editRow || this.editRows[this.changingRowIndex];
    };
    DynamicWizard.prototype.cancelRow = function () {
        var _this = this;
        var rowIndex = Utils._.clone(this.changingRowIndex);
        if (this.options.readOnly) {
            return;
        }
        var editRow = this.editRows[rowIndex];
        if ((editRow === null || editRow === void 0 ? void 0 : editRow.state) === FormState.New) {
            if (this.isSavingInProgress) {
                this.removeRow(rowIndex);
            }
            else {
                editRow.state = FormState.Removed;
                this.clearErrors(rowIndex);
                this.destroyComponents(rowIndex);
                if (this.inlineEditMode) {
                    this.splice(rowIndex);
                }
                this.editRows.splice(rowIndex, 1);
            }
        }
        else if ((editRow === null || editRow === void 0 ? void 0 : editRow.state) === FormState.Editing) {
            editRow.state = editRow.prevState;
            if (this.inlineEditMode) {
                this.dataValue[rowIndex] = editRow.backup;
            }
            editRow.data = editRow.backup;
            editRow.backup = null;
            if (this.isSavingInProgress) {
                this.setValues(editRow);
                this.rebuild(true);
            }
            this.restoreRowContext(editRow);
            if (!this.component.rowDrafts) {
                this.clearErrors(rowIndex);
            }
        }
        this.returnPrevPageState();
        this.shouldUpdate = true;
        this.emit('cancelRow');
        this.checkValidity(null, true);
        this.redraw().then(function () {
            _this.afterCancelRow();
        });
        if (this.component.rowDrafts) {
            this.checkValidity(this.data, false);
        }
    };
    DynamicWizard.prototype.afterCancelRow = function () {
        this.agreeButtonElements[0].focus();
    };
    DynamicWizard.prototype.rebuild = function (secondRender) {
        if (secondRender === void 0) { secondRender = false; }
        this.destroy();
        this.init(secondRender, this.prevVisibility);
        this.visible = this.conditionallyVisible(null, null);
        return this.redraw();
    };
    DynamicWizard.prototype.returnPrevPageState = function () {
        this.isChangingMode = false;
        this.isSavingInProgress = false;
        this.step = -1;
        this.changingRowIndex = null;
        this.removeChangingMode();
    };
    DynamicWizard.prototype.setValues = function (editRow) {
        if (this.options.readOnly) {
            return;
        }
        if (this.saveEditMode) {
            var dataValue = this.dataValue || [];
            if ((editRow === null || editRow === void 0 ? void 0 : editRow.state) === FormState.New && !this.isSavingInProgress) {
                var newIndex = dataValue.length;
                dataValue.push(editRow === null || editRow === void 0 ? void 0 : editRow.data);
                if (this.changingRowIndex !== newIndex) {
                    this.editRows.splice(this.changingRowIndex, 1);
                    this.editRows.splice(this.changingRowIndex, 0, editRow);
                }
            }
            else if ((editRow === null || editRow === void 0 ? void 0 : editRow.state) === FormState.Editing || this.isSavingInProgress) {
                dataValue[this.changingRowIndex] = editRow === null || editRow === void 0 ? void 0 : editRow.data;
            }
        }
    };
    DynamicWizard.prototype.saveRow = function (isRowValid) {
        var _this = this;
        var editRow = this.editRows[this.changingRowIndex];
        if (editRow) {
            this.setValues(editRow);
            if (this.isSavingInProgress) {
                this.isSavingInProgress = false;
            }
            editRow.state = this.component.rowDrafts && !isRowValid ? FormState.Draft : FormState.Saved;
            editRow.backup = null;
            this.updateValue();
            this.triggerChange();
            if (this.component.rowDrafts) {
                editRow.components.forEach(function (comp) { return comp.setPristine(_this.pristine); });
            }
            this.checkValidity(null, true);
            this.redraw().then(function () {
                _this.emit('saveRow');
                if (_this.afterSaveRow) {
                    _this.afterSaveRow();
                }
            });
            if (editRow.alerts) {
                editRow.alerts = false;
            }
            this.isChangingMode = false;
            this.changingRowIndex = null;
            this.shouldUpdate = true;
            this.removeChangingMode();
        }
        return true;
    };
    DynamicWizard.prototype.saveCurrentPageData = function (editRow) {
        if ((editRow === null || editRow === void 0 ? void 0 : editRow.state) === FormState.New || (editRow === null || editRow === void 0 ? void 0 : editRow.state) === FormState.Editing) {
            this.setValues(editRow);
            this.checkData(this.data);
            if (!this.isSavingInProgress) {
                this.isSavingInProgress = true;
            }
        }
    };
    DynamicWizard.prototype.isRowEditing = function (editRow) {
        return (editRow === null || editRow === void 0 ? void 0 : editRow.state) === FormState.Editing || (editRow === null || editRow === void 0 ? void 0 : editRow.state) === FormState.New;
    };
    DynamicWizard.prototype.switchToStep = function (component) {
        var relativePath = this.getRelativePath(component.path);
        var arrayPath = Utils.getArrayFromComponentPath(relativePath);
        var step = Utils._.findIndex(this.components, function (comp) { return comp.key === component.component.key; });
        var editRow = this.editRows[arrayPath[0]];
        var isAlreadyEditing = this.isRowEditing(editRow);
        if (!editRow) {
            this.addRow();
        }
        else if (!isAlreadyEditing) {
            this.editRow(arrayPath[0], step);
        }
        else {
            this.step = Utils._.isFinite(step) ? step : 0;
            this.addChangingMode();
            this.validateRow(editRow, false);
        }
    };
    DynamicWizard.prototype.editRow = function (rowIndex, step) {
        var editRow = this.editRows[rowIndex];
        var isAlreadyEditing = this.isRowEditing(editRow);
        if (!editRow || isAlreadyEditing) {
            return;
        }
        editRow.prevState = editRow.state;
        editRow.state = FormState.Editing;
        var dataSnapshot = Utils._.cloneDeep(editRow.data);
        if (this.inlineEditMode) {
            editRow.backup = dataSnapshot;
        }
        else {
            editRow.backup = editRow.data;
            editRow.data = dataSnapshot;
            this.restoreRowContext(editRow);
        }
        if (this.component.modal) {
            return this.addRowModal(rowIndex);
        }
        this.isChangingMode = true;
        this.changingRowIndex = rowIndex;
        this.step = Utils._.isFinite(step) ? step : 0;
        this.addChangingMode();
        this.validateRow(editRow, false);
    };
    DynamicWizard.prototype.removeRow = function (rowIndex) {
        var _this = this;
        if (this.options.readOnly) {
            return;
        }
        this.baseRemoveRow(rowIndex);
        this.splice(rowIndex);
        this.editRows.splice(rowIndex, 1);
        this.updateRowsComponents(rowIndex);
        this.updateValue();
        this.triggerChange();
        this.checkValidity(null, true);
        this.checkData();
        this.redraw().then(function () {
            _this.afterRemoveRow();
        });
        this.shouldUpdate = true;
    };
    DynamicWizard.prototype.afterRemoveRow = function () {
        this.agreeButtonElements[0].focus();
    };
    DynamicWizard.prototype.renderRow = function (row, rowIndex) {
        var _this = this;
        var dataValue = this.dataValue || [];
        var flattenedComponents = this.flattenComponents(rowIndex);
        var rowTemplate = Utils.Evaluator.noeval ? templates.row : Utils._.get(this.component, 'templates.row', DynamicWizard.defaultRowTemplate);
        return this.renderString(rowTemplate, {
            row: dataValue[rowIndex] || {},
            data: this.data,
            rowIndex: rowIndex,
            components: this.component.components,
            flattenedComponents: flattenedComponents,
            isVisibleInRow: function (component) { return _this.isComponentVisibleInRow(component, flattenedComponents); },
            getView: function (component, data) {
                var instance = flattenedComponents[component.key];
                var view = instance ? instance.getView(data || instance.dataValue) : '';
                if (instance && instance.widget && (view !== '--- PROTECTED ---')) {
                    if (Utils._.isArray(view)) {
                        view = view.map(function (value) { return instance.widget.getValueAsString(value); });
                    }
                    else {
                        view = instance.widget.getValueAsString(view);
                    }
                }
                return view;
            },
            state: this.editRows[rowIndex].state,
        });
    };
    DynamicWizard.prototype.hasButton = function (name, nextPage) {
        if (nextPage === void 0) { nextPage = this.getNextPage(); }
        // get page options with global options as default values
        var _a = this.buttonSettings, _b = _a.previous, previous = _b === void 0 ? this.buttonSettings.showPrevious : _b, _c = _a.cancel, cancel = _c === void 0 ? this.buttonSettings.showCancel : _c, _d = _a.next, next = _d === void 0 ? this.buttonSettings.showNext : _d;
        switch (name) {
            case 'previous':
                return previous && this.step !== 0 && this.hasComponents;
            case 'next':
                return next && nextPage && this.hasComponents;
            case 'cancel':
                return cancel && this.hasComponents;
            default:
                return true;
        }
    };
    DynamicWizard.prototype.getNextPage = function () {
        var _a;
        return this.step < ((_a = this.component) === null || _a === void 0 ? void 0 : _a.components.length);
    };
    DynamicWizard.prototype.prevPage = function () {
        var _a, _b;
        var editRow = this.editRows[this.changingRowIndex];
        if (this.step === 0) {
            this.rootWizard.prevPage();
            return;
        }
        else {
            var currStep = Utils._.clone(this.step);
            this.step = currStep - 1;
            if (!((_b = (_a = editRow === null || editRow === void 0 ? void 0 : editRow.components) === null || _a === void 0 ? void 0 : _a[this.step]) === null || _b === void 0 ? void 0 : _b._visible)) {
                this.prevPage();
                return;
            }
            this.saveCurrentPageData(editRow);
        }
    };
    DynamicWizard.prototype.findRootWizard = function (component) {
        var _a, _b, _c;
        var root = (_a = component === null || component === void 0 ? void 0 : component.parent) === null || _a === void 0 ? void 0 : _a.parent;
        if ((_b = component === null || component === void 0 ? void 0 : component.parent) === null || _b === void 0 ? void 0 : _b.subForm) {
            root = root === null || root === void 0 ? void 0 : root.parent;
        }
        return root && ((_c = root === null || root === void 0 ? void 0 : root._form) === null || _c === void 0 ? void 0 : _c.display) === 'wizard' ? this.findRootWizard(root) : component;
    };
    DynamicWizard.prototype.nextPage = function () {
        var _a, _b, _c, _d;
        var editRow = this.editRows[this.changingRowIndex];
        var isRowValid = this.validateRow(editRow, true);
        if (!this.component.rowDrafts) {
            if (!isRowValid) {
                return true;
            }
        }
        if (this.step === ((_b = (_a = this.component) === null || _a === void 0 ? void 0 : _a.components) === null || _b === void 0 ? void 0 : _b.length) - 1) {
            this.step = -1;
            this.saveRow(isRowValid);
            if (this.rootWizard.currentNextPage === -1) {
                this.rootWizard.redraw();
            }
            return false;
        }
        var prevStep = Utils._.clone(this.step);
        this.step = prevStep + 1;
        if (!((_d = (_c = editRow === null || editRow === void 0 ? void 0 : editRow.components) === null || _c === void 0 ? void 0 : _c[this.step]) === null || _d === void 0 ? void 0 : _d._visible)) {
            this.nextPage();
            return;
        }
        this.saveCurrentPageData(editRow);
    };
    DynamicWizard.prototype.validateStep = function (component, valid, dirty, rowData) {
        var isValid = valid;
        var hasServerError = !!(component.serverErrors && component.serverErrors.length);
        if (!this.component.rowDrafts && !hasServerError) {
            component.setPristine(!dirty);
        }
        isValid = isValid && component.checkValidity(null, dirty, rowData);
        return isValid;
    };
    ;
    DynamicWizard.prototype.validateRow = function (editRow, dirty) {
        var _this = this;
        var valid = true;
        if (!editRow) {
            return valid;
        }
        var errorsSnapshot = __spreadArrays(this.errors);
        if (editRow.state === FormState.Editing || dirty || (editRow.state === FormState.Draft && !this.pristine && !this.root.pristine)) {
            var component = editRow.components[this.step];
            if (component) {
                valid = this.validateStep(component, valid, dirty, editRow.data);
            }
            else if (this.step === -1) {
                editRow.components.forEach(function (comp) {
                    valid = _this.validateStep(comp, valid, dirty, editRow.data);
                });
            }
        }
        if (this.component.validate && this.component.validate.row) {
            valid = this.evaluate(this.component.validate.row, {
                valid: valid,
                row: editRow.data
            }, 'valid', true);
            if (valid.toString() !== 'true') {
                editRow.error = valid;
                valid = false;
            }
            else {
                editRow.error = null;
            }
            if (valid === null) {
                valid = "Invalid row validation for " + this.key;
            }
        }
        editRow.errors = !valid ? this.errors.filter(function (err) { return !errorsSnapshot.includes(err); }) : null;
        this.showRowErrorAlerts(editRow, !!valid);
        return !!valid;
    };
    DynamicWizard.prototype.findSourceRoot = function (root) {
        return (root === null || root === void 0 ? void 0 : root.parent) ? this.findSourceRoot(root === null || root === void 0 ? void 0 : root.root) : root;
    };
    DynamicWizard.prototype.changeState = function (changed) {
        var sourceRoot = this.root ? this.findSourceRoot(this.root) : {};
        if (changed && !sourceRoot.submissionInProcess) {
            !this.prevVisibility && this._visible ? this.rebuild(this.secondRender) : this.rebuild();
        }
        else {
            this.redraw();
        }
    };
    DynamicWizard.prototype.addChangingMode = function (firstPage) {
        var _this = this;
        if (firstPage === void 0) { firstPage = false; }
        if (this.rootWizard.element) {
            if (!this.rootWizard.element.classList.contains('dynamicWizard-changingMode')) {
                this.rootWizard.element.classList.add('dynamicWizard-changingMode');
            }
            if (!firstPage) {
                this.redraw().then(function () {
                    _this.afterAddChangingMode();
                });
                this.rootWizard.redraw();
            }
            this.triggerChange();
        }
    };
    DynamicWizard.prototype.afterAddChangingMode = function () {
        var _a, _b;
        var editRow = this.editRows[this.changingRowIndex];
        if ((_a = editRow.components[this.step].refs) === null || _a === void 0 ? void 0 : _a.input) {
            (_b = editRow.components[this.step].refs) === null || _b === void 0 ? void 0 : _b.input[0].focus();
        }
        else {
            this.focusOnNewRowElement(editRow.components, this.step);
        }
    };
    DynamicWizard.prototype.removeChangingMode = function () {
        var _this = this;
        var _a;
        if ((_a = this.rootWizard) === null || _a === void 0 ? void 0 : _a.element) {
            this.rootWizard.element.classList.remove('dynamicWizard-changingMode');
            this.redraw().then(function () {
                if (_this.afterRemoveChangingMode) {
                    _this.afterRemoveChangingMode();
                }
            });
            this.rootWizard.redraw();
            this.triggerChange();
        }
    };
    DynamicWizard.prototype.getValueAsString = function (value, options) {
        var _a;
        if (options === null || options === void 0 ? void 0 : options.email) {
            var result_1 = ("\n        <table border=\"1\" style=\"width:100%\">\n          <thead>\n            <tr>\n      ");
            (_a = this.component.components) === null || _a === void 0 ? void 0 : _a.forEach(function (component) {
                var label = component.label || component.key;
                result_1 += "<th style=\"padding: 5px 10px;\">" + label + "</th>";
            });
            result_1 += ("\n          </tr>\n        </thead>\n        <tbody>\n      ");
            this.iteratableRows.forEach(function (_a) {
                var components = _a.components;
                result_1 += '<tr>';
                _.each(components, function (component) {
                    result_1 += '<td style="padding:5px 10px;">';
                    if (component.isInputComponent && component.visible && !component.skipInEmail) {
                        result_1 += component.getView(component.dataValue, options);
                    }
                    result_1 += '</td>';
                });
                result_1 += '</tr>';
            });
            result_1 += ("\n          </tbody>\n        </table>\n      ");
            return result_1;
        }
        if (!value || !value.length) {
            return '';
        }
        return _super.prototype.getValueAsString.call(this, value, options);
    };
    DynamicWizard.editForm = editForm;
    return DynamicWizard;
}(EditGrid));
export default DynamicWizard;
