// @formio/premium/lib/components/index.js
import datasource from './DataSource/DataSource';
import sketchpad from './Sketchpad/Sketchpad';
import tagpad from './Tagpad/Tagpad';
import dynamicWizard from './DynamicWizard/DynamicWizard';
import datatable from './DataTable/DataTable';
import reviewPage from './ReviewPage/ReviewPage';
import premiumComponent from './Component/Component';
import checkbox from './Checkbox/Checkbox';
import datetime from './DateTime/DateTime';
import day from './Day/Day';
import file from './File/File';
import premiumSignature from './PremiumSignature/PremiumSignature';
import number from './Number/Number';
import radio from './Radio/Radio';
import select from './Select/Select';
import selectboxes from './SelectBoxes/SelectBoxes';
import survey from './Survey/Survey';
import tags from './Tags/Tags';
import textfield from './TextField/TextField';
export default {
    component: premiumComponent,
    base: premiumComponent,
    checkbox: checkbox,
    datetime: datetime,
    day: day,
    select: select,
    file: file,
    signature: premiumSignature,
    number: number,
    radio: radio,
    selectboxes: selectboxes,
    survey: survey,
    tags: tags,
    textfield: textfield,
    datasource: datasource,
    sketchpad: sketchpad,
    tagpad: tagpad,
    dynamicWizard: dynamicWizard,
    datatable: datatable,
    reviewpage: reviewPage
};
