// @formio/premium/lib/components/PremiumSignature/editForm/PremiumSignatureDisplay.edit.providers.js
import * as _ from 'lodash';
export default [
    {
        type: 'select',
        input: true,
        key: 'provider',
        label: 'Provider',
        placeholder: 'Select your esign storage provider',
        weight: 0,
        tooltip: 'Which signature component you want to use.',
        valueProperty: 'value',
        dataSrc: 'custom',
        template: '<span>{{ item.label }}</span>',
        defaultValue: {
            name: 'default',
            tagtype: 1
        },
        data: {
            custom: function () {
                return _.values(Formio.Providers.getProviders('esign')).sort().map(function (provider) { return ({
                    label: provider.name,
                    value: { name: provider.name, tagtype: provider.tagtype },
                }); });
            }
        },
    },
    {
        label: 'Signer Email',
        key: 'email',
        input: true,
        type: 'textfield',
        tooltip: 'An email of a person that needs to sign the document.',
        clearOnHide: true,
        validate: {
            custom: 'valid = /\\{\\{([\\s\\S]+?)\\}\\}/g.test(input) ? true : /^(([^<>()[\\]\\\\.,;:\\s@\"]+(\\.[^<>()[\\]\\\\.,;:\\s@\"]+)*)|(\".+\"))@((\\[[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$/.test(input)',
            customMessage: 'Must be a valid email',
            required: true
        },
        customConditional: "show = data['provider'].name && data['provider'].name !== 'default';",
        weight: 5
    },
    {
        label: 'Signer Order',
        key: 'order',
        input: true,
        type: 'number',
        tooltip: 'The order in which the person signs the document. Leaving this field blank will deliver the esign email to everyone all at once.',
        clearOnHide: true,
        customConditional: "show = data['provider'].name && data['provider'].name !== 'default';",
        weight: 5,
    },
];
